File: Project4/Complex.java
public class Complex {
    private double r;
    private double i;

    public Complex(double r, double i) {
        this.r = r;
        this.i = i;
    }

    public Complex add(Complex other) {
        return new Complex(this.r + other.r, this.i + other.i);
    }

    public Complex subtract(Complex other) {
        return new Complex(this.r - other.r, this.i - other.i);
    }

    public Complex multiply(Complex other) {
        return new Complex(this.r * other.r - this.i * other.i, this.r * other.i + this.i * other.r);
    }

    public Complex divide(Complex other) {
        double denom = other.r * other.r + other.i * other.i;
        return new Complex((this.r * other.r + this.i * other.i) / denom, (this.i * other.r - this.r * other.i) / denom);
    }

    @Override
    public String toString() {
        return String.format("(%.2f, %.2fj)", r, i);
    }

    public static void main(String[] args) {
        Complex a = new Complex(2, 3);
        Complex b = new Complex(1, -1);
        System.out.println(a + " + " + b + " = " + a.add(b));
        System.out.println(a + " - " + b + " = " + a.subtract(b));
        System.out.println(a + " * " + b + " = " + a.multiply(b));
        System.out.println(a + " / " + b + " = " + a.divide(b));
    }
}

File: Project4/complex.py
class Complex:
    def __init__(self, num=0, den=1):
        self.num = num
        self.den = den

    def __add__(self, other):
        if isinstance(other, (int, float)):
            return Complex(self.num + other, self.den)
        elif isinstance(other, Complex):
            return Complex(self.num + other.num, self.den + other.den)
        else:
            raise TypeError

    def __sub__(self, other):
        if isinstance(other, (int, float)):
            return Complex(self.num - other, self.den)
        elif isinstance(other, Complex):
            return Complex(self.num - other.num, self.den - other.den)
        else:
            raise TypeError

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            return Complex(self.num * other, self.den * other)
        elif isinstance(other, Complex):
            return Complex(self.num * other.num - self.den * other.den,
                           self.num * other.den + self.den * other.num)
        else:
            raise TypeError

    def __truediv__(self, other):
        if isinstance(other, (int, float)):
            return Complex(self.num / other, self.den / other)
        elif isinstance(other, Complex):
            denom = other.num ** 2 + other.den ** 2
            return Complex((self.num * other.num + self.den * other.den) / denom,
                           (self.den * other.num - self.num * other.den) / denom)
        else:
            raise TypeError

    def __str__(self):
        return f"({self.num:.2f} + {self.den:.2f}j)"

if __name__ == "__main__":
    a = Complex(2.0, 3.0)
    b = Complex(1.0, -1.0)

    print(f"{a} + {b} = {a + b}")
    print(f"{a} - {b} = {a - b}")
    print(f"{a} * {b} = {a * b}")
    print(f"{a} / {b} = {a / b}")
